src/treemodel/schema.js

```diff
@@ -0,0 +1,417 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import clone from '../lib/lodash/clone.js';
+import CKEditorError from '../ckeditorerror.js';
+
+/**
+ * SchemaItem is a singular registry item in {@link core.treeModel.Schema} that groups and holds allow/disallow rules for
+ * one entity. This class is used internally in {@link core.treeModel.Schema} and should not be used outside it.
+ *
+ * @see core.treeModel.Schema
+ * @memberOf core.treeModel
+ */
+export class SchemaItem {
+	/**
+	 * Creates SchemaItem instance.
+	 *
+	 * @param {core.treeModel.Schema} schema Schema instance that owns this item.
+	 */
+	constructor( schema ) {
+		/**
+		 * Schema instance that owns this item.
+		 *
+		 * @member {core.treeModel.Schema} core.treeModel.SchemaItem#_schema
+		 * @private
+		 */
+		this._schema = schema;
+
+		/**
+		 * Paths in which the entity, represented by this item, is allowed.
+		 *
+		 * @member {Array} core.treeModel.SchemaItem#_allowed
+		 * @private
+		 */
+		this._allowed = [];
+
+		/**
+		 * Paths in which the entity, represented by this item, is disallowed.
+		 *
+		 * @member {Array} core.treeModel.SchemaItem#_disallowed
+		 * @private
+		 */
+		this._disallowed = [];
+	}
+
+	/**
+	 * Allows entity, represented by this item, to be in given path.
+	 *
+	 * @param {Array.<String>|String} path Path in which entity is allowed. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 */
+	addAllowed( path, attribute ) {
+		this._addPath( '_allowed', path, attribute );
+	}
+
+	/**
+	 * Disallows entity, represented by this item, to be in given path.
+	 *
+	 * @param {Array.<String>|String} path Path in which entity is disallowed. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 */
+	addDisallowed( path, attribute ) {
+		this._addPath( '_disallowed', path, attribute );
+	}
+
+	/**
+	 * Adds path to the SchemaItem instance.
+	 *
+	 * @param {String} member Name of the array member into which the path will be added. Possible values are `_allowed` or `_disallowed`.
+	 * @param {Array.<String>|String} path Path to be added. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 * @private
+	 */
+	_addPath( member, path, attribute ) {
+		if ( typeof path === 'string' ) {
+			path = path.split( ' ' );
+		}
+
+		path = path.slice();
```

Comment:
As far as `_getPaths` is concerned -- we use that function in `_hasMatchingPath` which shifts `itemPath` array. If we won't copy the array in `_getPaths` we will be shifting the original array that is stored as class member. We need a copy of that array because we use it for temporary processing.